secboot/secboot_hooks.go

```diff
@@ -135,30 +142,41 @@ func ResealKeysWithFDESetupHook(keyFiles []string, primaryKeyFile string, models
 	for _, model := range models {
 		sbModels = append(sbModels, model)
 	}
-	for _, keyFile := range keyFiles {
-		loadedKey := &defaultKeyLoader{}
-		const hintExpectFDEHook = true
-		if err := readKeyFile(keyFile, loadedKey, hintExpectFDEHook); err != nil {
-			return err
-		}
+	for _, key := range keys {
+		var keyDataWriter sb.KeyDataWriter
 
-		// Non-nil FDEHookKeyV1 indicates that V1 hook key is used
-		if loadedKey.FDEHookKeyV1 != nil {
-			// V1 keys do not need resealing
-			continue
-		}
+		keyData, tokenWriter, tokenErr := key.readTokenAndGetWriter()
+		if tokenErr == nil {
+			keyDataWriter = tokenWriter
+		} else {
+			loadedKey := &defaultKeyLoader{}
+			const hintExpectFDEHook = true
+			if err := readKeyFile(key.KeyFile, loadedKey, hintExpectFDEHook); err != nil {
+				return tokenErr
+			}
+
+			// Non-nil FDEHookKeyV1 indicates that V1 hook key is used
+			if loadedKey.FDEHookKeyV1 != nil {
+				// V1 keys do not need resealing
+				continue
+			}
+
+			if loadedKey.KeyData == nil {
+				return fmt.Errorf("internal error: keydata was expected, but none found")
+			}
+
+			keyData = loadedKey.KeyData
```

Comment:
Nitpick, This part could be extracted into `KeyDataLocation.readTokenAndGetWriter` which then could be renamed to `KeyDataLocation.readKeyDataAndGetWriter`.

Reply:
Actually, It could be simpler to extend `readKeyData` to take hook hint and reuse it here instead.

Reply:
It gets messy. Specially on how to handle if the key was a v1 hook key. Do I return another type? There is already 4 returns to `readKeyData`. Or do I say that if there is no KeyData and no SealedKeyObject, then it has to be a v1 hook key? Or do I use a special error? Anyway I try to change this to factor it, gets the factored function more difficult without context, but we still have to sort out the type of return in the caller. Which just makes more code.